src/app/auth/role-guard.service.ts

```diff
@@ -0,0 +1,28 @@
+import { Injectable } from '@angular/core';
+import { 
+  Router,
+  CanActivate,
+  ActivatedRouteSnapshot
+} from '@angular/router';
+import { AuthService } from './auth.service';
+import decode from 'jwt-decode';
+@Injectable()
+export class RoleGuardService implements CanActivate {
+  constructor(public auth: AuthService, public router: Router) {}
+  canActivate(route: ActivatedRouteSnapshot): boolean {
+    // this will be passed from the route config
+    // on the data property
+    const expectedRole = route.data.expectedRole;
+    const token = localStorage.getItem('token');
+    // decode the token to get its payload
```

Comment:
I would advice not to use local storage in guards.

Reply:
I have revamped the entire structure keeping in mind about your points. Please check!